mesonbuild/ast/interpreter.py

```
@@ -215,45 +230,91 @@ def evaluate_foreach(self, node):
-        if isinstance(args, mparser.ArrayNode):
+
+        if isinstance(args, ArrayNode):
             args = [x for x in args.args.arguments]
-        elif isinstance(args, mparser.ArgumentNode):
+
+        elif isinstance(args, ArgumentNode):
             args = [x for x in args.arguments]
-        for i in args:
-            if isinstance(i, mparser.ArrayNode):
-                temp_args += [x for x in i.args.arguments]
+
+        elif isinstance(args, ArithmeticNode):
+            if args.operation != 'add':
+                return [] # Only handle string and array concats
+            l = quick_resolve(args.left)
+            r = quick_resolve(args.right)
+            if isinstance(l, str) and isinstance(r, str):
+                args = [l + r] # String concatination detected
             else:
-                temp_args += [i]
-        for i in temp_args:
-            if isinstance(i, mparser.ElementaryNode) and not isinstance(i, mparser.IdNode):
+                args = self.flatten_args(l, include_unknown_args, id_loop_detect) + self.flatten_args(r, include_unknown_args, id_loop_detect)
+
+        elif isinstance(args, MethodNode):
+            src = quick_resolve(args.source_object)
+            margs = self.flatten_args(args.args, include_unknown_args, id_loop_detect)
+            try:
+                if isinstance(src, str):
+                    args = [self.string_method_call(src, args.name, margs)]
+                elif isinstance(src, bool):
+                    args = [self.bool_method_call(src, args.name, margs)]
+                elif isinstance(src, int):
+                    args = [self.int_method_call(src, args.name, margs)]
+                elif isinstance(src, list):
+                    args = [self.array_method_call(src, args.name, margs)]
+                elif isinstance(src, dict):
+                    args = [self.dict_method_call(src, args.name, margs)]
+                else:
+                    return []
+            except Exception:
```

I'd prefer to narrow this except if possible.

---

I changed it to `MesonException`. I wouldn't go for a more specific class than that, so that all meson errors are ignored.